<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://static.vecteezy.com/system/resources/previews/001/349/456/non_2x/futuristic-background-with-polygonal-lines-free-vector.jpg');color:white;font-size:16px">
The Projects goal is to identify the best selling books and authors along with which books got the highest reviews, while excluding brochures and similar publications.

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp',             # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,                        # connection port
             'db': 'data-analyst-final-project-db'} # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://static.vecteezy.com/system/resources/previews/001/349/456/non_2x/futuristic-background-with-polygonal-lines-free-vector.jpg');color:white; font-size:48px">
<center><b>Study the tables</b></center>

<center><b>BOOKS</b></center>

In [2]:
books = '''
Select *
From books

'''


pd.io.sql.read_sql(books, con = engine).head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
pd.io.sql.read_sql(books, con = engine).shape

(1000, 6)

<center><b>AUTHORS</b></center>

In [4]:
authors = '''
Select *
From authors

'''


pd.io.sql.read_sql(authors, con = engine).head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
pd.io.sql.read_sql(authors, con = engine).shape

(636, 2)

<center><b>RATINGS</b></center>

In [6]:
ratings = '''
Select *
From ratings

'''


pd.io.sql.read_sql(ratings, con = engine).head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
pd.io.sql.read_sql(ratings, con = engine).shape

(6456, 4)

<center><b>REVIEWS</b></center>

In [8]:
reviews = '''
Select *
From reviews

'''


pd.io.sql.read_sql(reviews, con = engine).head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [9]:
pd.io.sql.read_sql(reviews, con = engine).shape

(2793, 4)

<center><b>PUBLISHERS</b></center>

In [10]:
publishers = '''
Select *
From publishers

'''


pd.io.sql.read_sql(publishers, con = engine).head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [11]:
pd.io.sql.read_sql(publishers, con = engine).shape

(340, 2)

<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://static.vecteezy.com/system/resources/previews/001/349/456/non_2x/futuristic-background-with-polygonal-lines-free-vector.jpg');color:white; font-size:20px">
<center><b>Find the number of books released after January 1, 2000.</b></center>

In [12]:
years = '''
SELECT
    min(publication_date) AS min,
    max(publication_date) AS max

FROM
    books
    

    
'''


pd.io.sql.read_sql(years, con = engine)

min         max
0  1952-12-01  2020-03-31

In [13]:
query_1 = '''
SELECT
    count(book_id) 

FROM
    books
    
WHERE
    publication_date::date >= '2000-01-01'
    
'''


pd.io.sql.read_sql(query_1, con = engine)

count
0    821

<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://t4.ftcdn.net/jpg/02/19/81/13/360_F_219811319_NJIHRK2sKH0SHnMBOf0il8KCcxKU2UuF.jpg');color:white; font-size:14px">
The dates range from end of 1952 up to March, 2020.<br>
The number of books released after January 1,2000 is 821.

<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://static.vecteezy.com/system/resources/previews/001/349/456/non_2x/futuristic-background-with-polygonal-lines-free-vector.jpg');color:white; font-size:20px">
<center><b>Find the number of user reviews and the average rating for each book.</b></center>

In [14]:
query_2 = '''
SELECT
    *
FROM
    ratings
    INNER JOIN reviews ON reviews.book_id = ratings.book_id
'''


pd.io.sql.read_sql(query_2, con = engine)

rating_id  book_id       username  rating  review_id  book_id  \
0              1        1     ryanfranco       4          2        1   
1              1        1     ryanfranco       4          1        1   
2              2        1  grantpatricia       2          2        1   
3              2        1  grantpatricia       2          1        1   
4              3        1   brandtandrea       5          2        1   
...          ...      ...            ...     ...        ...      ...   
26162       6455     1000          zross       2       2790     1000   
26163       6456     1000        fharris       5       2793     1000   
26164       6456     1000        fharris       5       2792     1000   
26165       6456     1000        fharris       5       2791     1000   
26166       6456     1000        fharris       5       2790     1000   

             username                                               text  
0          ryanfranco  Foot glass pretty audience hit themselves. Amo...  
1        brandtandrea  Mention society tell send professor analysis. ...  
2          ryanfranco  Foot glass pretty audience hit themselves. Amo...  
3        brandtandrea  Mention society tell send professor analysis. ...  
4          ryanfranco  Foot glass pretty audience hit themselves. Amo...  
...               ...                                                ...  
26162         wknight   Change lose answer close pressure. Spend so now.  
26163      jarvispaul  Republican staff bit eat material measure plan...  
26164         wendy18  Or western offer wonder ask. More hear phone f...  
26165  carolrodriguez  Authority go who television entire hair guy po...  
26166         wknight   Change lose answer close pressure. Spend so now.  

[26167 rows x 8 columns]

In [15]:
query_2 = '''
SELECT
   ratings.book_id,
   COUNT(DISTINCT(review_id)) AS review_cnt,
   AVG(rating)

FROM
    ratings
    INNER JOIN reviews ON reviews.book_id = ratings.book_id
    
GROUP BY
    ratings.book_id
    
ORDER BY
    review_cnt DESC
'''


pd.io.sql.read_sql(query_2, con = engine)

book_id  review_cnt       avg
0        948           7  3.662500
1        854           6  3.772727
2        656           6  4.264151
3        734           6  4.206897
4        963           6  3.977273
..       ...         ...       ...
989      465           1  3.500000
990      446           1  3.000000
991      431           1  4.000000
992       92           1  3.500000
993      413           1  4.000000

[994 rows x 3 columns]

<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://t4.ftcdn.net/jpg/02/19/81/13/360_F_219811319_NJIHRK2sKH0SHnMBOf0il8KCcxKU2UuF.jpg');color:white; font-size:14px">
In the data we have, Reviews have total of 2,793 entries, and the number of the highest review count is 7.<br>
This data represents books that are spanning for 68 year! Although the source of the data is unknown and we dont have dates of reviews.

<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://static.vecteezy.com/system/resources/previews/001/349/456/non_2x/futuristic-background-with-polygonal-lines-free-vector.jpg');color:white; font-size:20px">
<center><b>Identify the publisher that has released the greatest number of books with more than 50 pages</b></center>

In [16]:
query_3 = '''
SELECT
    books.publisher_id,
    publishers.publisher,
    count(books.book_id)

FROM
   books
   INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
    
WHERE
    books.num_pages > 50
GROUP BY
    books.publisher_id,
    publishers.publisher
ORDER BY
    count DESC
'''


pd.io.sql.read_sql(query_3, con = engine)

publisher_id                              publisher  count
0             212                          Penguin Books     42
1             309                                Vintage     31
2             116               Grand Central Publishing     25
3             217                       Penguin Classics     24
4              33                       Ballantine Books     19
..            ...                                    ...    ...
329           148          Harvard Business Review Press      1
330           302               Tyndale House Publishers      1
331            66                        Corgi Childrens      1
332           230  Pocket Books / Simon & Schuster  Inc.      1
333           199                          Orchard Books      1

[334 rows x 3 columns]

<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://t4.ftcdn.net/jpg/02/19/81/13/360_F_219811319_NJIHRK2sKH0SHnMBOf0il8KCcxKU2UuF.jpg');color:white; font-size:14px">
"Penguin Books" released the highest amount of books (at 42 total).

<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://static.vecteezy.com/system/resources/previews/001/349/456/non_2x/futuristic-background-with-polygonal-lines-free-vector.jpg');color:white; font-size:20px">
<center><b>Identify the author with the highest average book rating (look only at books with at least 50 ratings).</b></center>

In [17]:
query_4 = '''
SELECT
    sub.author_id,
    sub.author,
    sub.avg,
    sub.cnt

FROM
    (SELECT 
        books.author_id,
        authors.author,
        AVG(rating) AS avg,
        COUNT(rating_id) AS cnt
    FROM 
        books
        INNER JOIN authors ON authors.author_id = books.author_id
        RIGHT JOIN ratings ON ratings.book_id = books.book_id
    GROUP BY
        books.author_id,
        authors.author
        ) AS sub
WHERE
    cnt > 50
ORDER BY
    avg DESC



'''


pd.io.sql.read_sql(query_4, con = engine)

author_id                                             author       avg  \
0         236                         J.K. Rowling/Mary GrandPré  4.288462   
1           3                                    Agatha Christie  4.283019   
2         402                  Markus Zusak/Cao Xuân Việt Khương  4.264151   
3         240                                     J.R.R. Tolkien  4.240964   
4         499                           Roald Dahl/Quentin Blake  4.209677   
5         376                                  Louisa May Alcott  4.203704   
6         498                                       Rick Riordan  4.130952   
7          39                                      Arthur Golden  4.107143   
8         542                                       Stephen King  4.009434   
9         302                                       John Grisham  3.971429   
10        621                                    William Golding  3.901408   
11        445                                    Nicholas Sparks  3.882883   
12        291                                       Jodi Picoult  3.881579   
13        536                                    Sophie Kinsella  3.877193   
14        257                                    James Patterson  3.859375   
15        235                                      J.D. Salinger  3.846939   
16        469           Paulo Coelho/Alan R. Clarke/Özdemir İnce  3.789474   
17        630  William Shakespeare/Paul Werstine/Barbara A. M...  3.787879   
18        106                                          Dan Brown  3.741259   
19        372                                         Lois Lowry  3.738462   
20        195            George Orwell/Boris Grabnar/Peter Škerl  3.729730   
21        554                                    Stephenie Meyer  3.662500   
22        311                                     John Steinbeck  3.643836   

    cnt  
0   312  
1    53  
2    53  
3   166  
4    62  
5    54  
6    84  
7    56  
8   106  
9    70  
10   71  
11  111  
12   76  
13   57  
14   64  
15   98  
16   57  
17   66  
18  143  
19   65  
20   74  
21  160  
22   73

<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://t4.ftcdn.net/jpg/02/19/81/13/360_F_219811319_NJIHRK2sKH0SHnMBOf0il8KCcxKU2UuF.jpg');color:white; font-size:14px">
The author with the highest book rating is "J.K. Rowling/Mary GrandPré", and also with the most reviews overall. Who doesnt love harry potter 🤩.

<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://static.vecteezy.com/system/resources/previews/001/349/456/non_2x/futuristic-background-with-polygonal-lines-free-vector.jpg');color:white; font-size:20px">
<center><b>Find the average number of text reviews among users who rated more than 50 books.</b></center>

In [18]:
query_5 = '''
SELECT
    AVG(reviews_count.text_count)
FROM
    (SELECT
        reviews.username,
        COUNT(reviews.text) AS text_count
    FROM
        reviews
    WHERE
        username IN (SELECT
            ratings.username
        FROM
            ratings
        GROUP BY
            ratings.username
        HAVING
            COUNT(book_id) >= 50)
    GROUP BY
            reviews.username) AS reviews_count
'''


pd.io.sql.read_sql(query_5, con = engine)

avg
0  24.222222

<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://t4.ftcdn.net/jpg/02/19/81/13/360_F_219811319_NJIHRK2sKH0SHnMBOf0il8KCcxKU2UuF.jpg');color:white; font-size:14px">
The average number of texts for users who rated more than 50 books is 24.22 

<div style="border:inset black 1px; border-radius: 5px; padding: 20px;background-image: url('https://t4.ftcdn.net/jpg/02/19/81/13/360_F_219811319_NJIHRK2sKH0SHnMBOf0il8KCcxKU2UuF.jpg');color:white; font-size:14px">
From the year 2000 only 821 books were released(thats very little for 20 years), books seem to be getting only up to 7 reviews with average ratings ranging between 3 and 4 score.<br>
Penguin Books publisher seems to publish the biggest amount of books by now (42 in count), seems like J.K. Rowling/Mary GrandPré has the most reviews and the highest rating average (Did i mention Harry Potter before? also spanning in the course of 7 books!).<br>
And lastly the hardcore book raters on average give 24 text reviews.
(Pretty much seems like books are dying out...).